File: app/services/document_names.py
"""Sanitize and display document filenames (storage may use UUID paths)."""

from __future__ import annotations

import re
from pathlib import Path

_UUID_FILENAME = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
    r"\.(pdf|txt|md)$",
    re.IGNORECASE,
)

_FRIENDLY_BY_EXT = {
    "pdf": "Uploaded PDF",
    "txt": "Uploaded text",
    "md": "Uploaded markdown",
}


def sanitize_upload_filename(filename: str | None) -> str:
    """Keep only the basename so path junk never lands in Document.filename."""
    name = Path(filename or "upload").name.strip() or "upload"
    return name[:255]


def display_document_name(filename: str | None) -> str:
    """Human label for UI/citations when the stored name is a bare UUID file."""
    name = (filename or "").strip() or "upload"
    match = _UUID_FILENAME.match(name)
    if not match:
        return name
    ext = match.group(1).lower()
    return _FRIENDLY_BY_EXT.get(ext, f"Uploaded {ext}")

File: frontend/src/documentNames.ts
const UUID_FILENAME =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\.(pdf|txt|md)$/i;

const FRIENDLY_BY_EXT: Record<string, string> = {
  pdf: "Uploaded PDF",
  txt: "Uploaded text",
  md: "Uploaded markdown",
};

/** Prefer a human label when a citation/doc name is a bare storage UUID. */
export function displayDocumentName(filename: string | null | undefined): string {
  const name = (filename || "").trim() || "upload";
  const match = UUID_FILENAME.exec(name);
  if (!match) {
    return name;
  }
  const ext = match[1].toLowerCase();
  return FRIENDLY_BY_EXT[ext] || `Uploaded ${ext}`;
}

/** Chat title from an upload: basename without the final extension. */
export function chatTitleFromFilename(filename: string | null | undefined): string {
  const base = displayDocumentName(filename).split(/[/\\]/).pop()?.trim() || "upload";
  const stem = base.replace(/\.[^.]+$/, "").trim();
  return (stem || base).slice(0, 200);
}
